### PR TITLE
Disable PKCE for confidential clients

### PR DESCRIPTION
**Pull request by danj3**
I found this necessary when using oidcc_plug to enable generation auth urls using confidential clients. If there was another approach, it was not apparent to me. Thank you for writing oidcc, a vast improvement over the ad-hoc combination of OAuth2 and Keycloak projects I was using.

### CODE REVIEW

**Comment by paulswartz** [comment]
Confidential clients can (and should) still use PKCE. Were you working with a client which doesn't validate when PKCE is used?

**Comment by maennchen** [comment]
I've used OpenID providers before that supported (and even suggested) using PKCE for confidential clients. But I've also seen multiple implementations doing what this PR does.

The RFC mentions this about the topic:

> OAuth 2.0 [[RFC6749](https://datatracker.ietf.org/doc/html/rfc6749)] public clients are susceptible to the
>   authorization code interception attack.
>
>   In this attack, the attacker intercepts the authorization code
>   returned from the authorization endpoint within a communication path
>   not protected by Transport Layer Security (TLS), such as inter-
>   application communication within the client's operating system.
> [...]

https://datatracker.ietf.org/doc/html/rfc7636#section-1

While this suggests that the RFC is designed for public clients, it never specifically restricts it to public clients.

File: mix.exs [hunk 1]
```diff
@@ -4,7 +4,7 @@ defmodule Oidcc.Plug.MixProject do
   def project do
     [
       app: :oidcc_plug,
-      version: "0.1.1",
```
Comment (maennchen) [inline mix.exs:7]: Releases are handled outside of PRs.

**Comment by paulswartz** [comment]
It's also RECOMMENDED by OAuth2 Security Best Current Practice: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#section-2.1.1
> For confidential clients, the use of PKCE [RFC7636] is RECOMMENDED, as it provides a strong protection against misuse and injection of authorization codes as described in Section 4.5.3.1 and, as a side-effect, prevents CSRF even in presence of strong attackers as described in Section 4.7.1.

If you want to disable PKCE, I believe adding a configuration override to set `code_challenge_methods_supported` to `[]` might work?

**Comment by maennchen** [comment]
@paulswartz Oh, good to know :+1: 

In that case I tend to agree.

**Comment by danj3** [comment]
I agree, the coupling of a confidential client with enable/disable PKCE is not correct, the option should be specific to PKCE. At the time I was unsuccessful combining a confidential client and PKCE with Keycloak. Your comments reinforce that this must be possible, so I will revisit and follow up.

**Comment by paulswartz** [comment]
If you continue to run into issues, feel free to check in on Slack: I'm @paulswartz on both the Elixir and EFF instances.

**Comment by danj3** [comment]
I've discovered my specific problem and it appears to be a bug in Keycloak when the code_verifier string contains URL escaped chars, like %2F for /, it fails. To isolate I changed to base32 and reduced the crypt string  to keep the length < 128 and then it started working. Are any of these properties in question from a standards perspective?

**Comment by danj3** [comment]
main fixes this problem completely!

**Comment by maennchen** [comment]
Awesome! I‘ll close this PR then.

I‘m a bit busy at the moment. (Traveling to the Elixir Conf today 😁) I‘ll get a release out next week.

**Comment by maennchen** [comment]
@danj3 Fix is released as `v0.1.2`.

It was great meeting you at Elixir Conf :smile:

**Pull request closed without merging**